_source/_posts/2021-08-13-reactive-java.md

```diff
@@ -0,0 +1,787 @@
+2021-07-15 22:57:33.436 - INFO [parallel-1] - element "5" [subscribe1]
+2021-07-15 22:57:33.436 - INFO [parallel-2] - element "5" [subscribe2]
+2021-07-15 22:57:33.438 - INFO [parallel-3] - element "1" [map]
+2021-07-15 22:57:33.438 - INFO [parallel-3] - element "1" [subscribe3]
+2021-07-15 22:57:33.439 - INFO [parallel-3] - element "2" [map]
+2021-07-15 22:57:33.439 - INFO [parallel-3] - element "2" [subscribe3]
+2021-07-15 22:57:33.439 - INFO [parallel-3] - element "3" [map]
+2021-07-15 22:57:33.439 - INFO [parallel-3] - element "3" [subscribe3]
+2021-07-15 22:57:33.441 - INFO [parallel-4] - element "1" [map]
+2021-07-15 22:57:33.441 - INFO [parallel-3] - element "4" [map]
+2021-07-15 22:57:33.442 - INFO [parallel-4] - element "1" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "1" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "2" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "2" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "1" [subscribe5]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "3" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "3" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "2" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "4" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "2" [subscribe5]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "4" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "3" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "5" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "5" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "3" [subscribe5]
+2021-07-15 22:57:33.442 - INFO [parallel-3] - element "4" [subscribe3]
+2021-07-15 22:57:33.450 - INFO [parallel-3] - element "5" [map]
+2021-07-15 22:57:33.450 - INFO [parallel-3] - element "5" [subscribe3]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "4" [map]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "4" [subscribe5]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "5" [map]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "5" [subscribe5]
+```
+
+As you can see, different subscription executions interleave, and as I am testing with only four CPU cores, four workers seem to be available: _parallel-1_, _parallel-2_, _parallel-3_, and _parallel-4_. The `subscription5` operations, execute in _parallel-1_.
+
+Again, observe how all operations for a given subscription are executed in the same thread.
+
+### Using Both `publishOn` and `subscribeOn`
+
+When the `publishOn` and `subscribeOn` operators are used, the `subscribe` call accepts a consumer. The consumer will execute in the context selected by `publishOn`, which defines the scheduler for downstream processing, including the consumer execution. At first this might be confusing to understand, so check the following example:
```

Comment:
```suggestion
When the `publishOn()` and `subscribeOn()` operators are used, the `subscribe` call accepts a consumer. The consumer will execute in the context selected by `publishOn()`, which defines the scheduler for downstream processing, including the consumer execution. At first, this might be confusing to understand, so check the following example:
```